### DISCUSSION TITLE
Read Replica Support

### DISCUSSION THREAD

**Discussion by Flo4604** [stephenafamo/bob#200] [Ideas]
It would be nice to have read replica support out of the box, by passing multiple sql.db's and saying whether they are the master / read replica

**Comment by stephenafamo** [comment]
Since every function/method asks for a `bob.Executor` this wouldn't really be Bob's responsibility. I assume it would be possible to create an `*sql.DB` that understands the primary/replica routing and use that.

I'm not sure how other implementations of something like this look like, can you share?

**Reply by Flo4604** [reply]
There is https://github.com/bxcodec/dbresolver

**Reply by stephenafamo** [reply]
Because **Bob** only needs a valid executor, you can use that package.


```go
	connectionDB := dbresolver.New(...)
	defer connectionDB.Close()

	bobExecutor := bob.New(connectionDB)
```

Does this work for your use case?

**Reply by Flo4604** [reply]
The packages has its own types that should be based of sql.DB and its other types but still imeplements most if it itself to support the multiple connections. 
So bob.New() doesn't like the type
```go
T (type dbresolver.DB) does not satisfy bob.StdInterface (wrong type for method PrepareContext)
		have PrepareContext(context.Context, string) (dbresolver.Stmt, error)
		want PrepareContext(context.Context, string) (*sql.Stmt, error)
```

I will try to edit the dbresolver package so it satisfies sql.db so I can pass it to bob.NewDB()

**Reply by stephenafamo** [reply]
It seems you would have to write a wrapper for it, similar to the wrappers **Bob** has for the standard library types:

* `*sql.DB`
* `*sql.Tx`
* `*sql.Conn`

Take a look at the `stdlib.go`

**Reply by Flo4604** [reply]
thanks for the pointers i'll look into this
